File: apps/web/scripts/audit_professors.py
import json
import re

with open('C:/Users/huama/.gemini/antigravity/brain/39f2b139-4ace-4a87-bd39-b1176155fd28/.system_generated/steps/2578/output.txt', 'r', encoding='utf-8') as f:
    raw = f.read()

# regex to find id, nombre, especialidad
pattern = r'{"id":"([^"]+)","nombre":"([^"]+)","especialidad":"([^"]+)"'
# taking into account some names might have backslashes \"
pattern = r'{"id":"([0-9a-f-]+)","nombre":"(.*?)(?<!\\)","especialidad":"(.*?)(?<!\\)"'
matches = re.findall(pattern, raw)

profs = []
for m in matches:
    profs.append({
        'id': m[0],
        'nombre': m[1].replace('\\"', ''),
        'especialidad': m[2]
    })

def get_words(name):
    # Remove accents, punctuation, etc.
    name = re.sub(r'[".,-]', ' ', name.lower())
    words = [w for w in name.split() if len(w) > 2 and w not in ['del', 'las', 'los', 'san', 'de']]
    return set(words)

duplicates = []
processed = set()

for i, p1 in enumerate(profs):
    if p1['id'] in processed: continue
    
    words1 = get_words(p1['nombre'])
    matches = [p1]

    for j, p2 in enumerate(profs[i+1:]):
        if p2['id'] in processed: continue
        
        words2 = get_words(p2['nombre'])
        intersect = words1.intersection(words2)
        
        if len(intersect) >= 2 and len(intersect) == len(words1) and len(intersect) == len(words2):
            matches.append(p2)
        elif len(intersect) >= 3 and (len(intersect) / len(words1.union(words2))) >= 0.75:
            matches.append(p2)
        elif len(intersect) == 3 and len(words1) == 3 and len(words2) == 3:
            matches.append(p2)

    if len(matches) > 1:
        duplicates.append(matches)
        for m in matches:
            processed.add(m['id'])

print(f"Found {len(duplicates)} potential duplicate groups.")
for i, group in enumerate(duplicates):
    print(f"\nGroup {i+1}:")
    for p in group:
        print(f"  - {p['nombre']} (ID: {p['id']}) [Esp: {p['especialidad']}]")

File: apps/web/scripts/audit_professors.ts
import fs from 'fs';
import path from 'path';

function normalizeName(name: string) {
  return name.normalize("NFD").replace(/[\u0300-\u036f]/g, "").replace(/[,.-]/g, " ").toLowerCase().trim();
}

function getWords(name: string) {
  const words = normalizeName(name).split(/\s+/).filter(w => w.length > 2 && w !== 'del' && w !== 'las' && w !== 'los' && w !== 'san');
  return new Set(words);
}

function intersection(setA: Set<string>, setB: Set<string>) {
  let _intersection = new Set();
  for (let elem of setB) {
      if (setA.has(elem)) {
          _intersection.add(elem);
      }
  }
  return _intersection;
}

async function run() {
  const filePath = path.resolve('C:/Users/huama/.gemini/antigravity/brain/39f2b139-4ace-4a87-bd39-b1176155fd28/.system_generated/steps/2578/output.txt');
  const rawOutput = fs.readFileSync(filePath, 'utf8');
  const startIdx = rawOutput.indexOf('[');
  const endIdx = rawOutput.lastIndexOf(']');
  const jsonStr = rawOutput.slice(startIdx, endIdx + 1);
  const profs = JSON.parse(jsonStr);

  const duplicates: any[] = [];
  const processed = new Set();

  for (let i = 0; i < profs.length; i++) {
    const p1 = profs[i];
    if (processed.has(p1.id)) continue;
    
    const words1 = getWords(p1.nombre);
    const matches = [p1];

    for (let j = i + 1; j < profs.length; j++) {
      const p2 = profs[j];
      if (processed.has(p2.id)) continue;

      const words2 = getWords(p2.nombre);
      const intersect = intersection(words1, words2);
      
      const unionSize = new Set([...words1, ...words2]).size;
      const matchRatio = intersect.size / unionSize;

      // Match if they share at least 2 distinct words and are exact,
      // or if they overlap highly.
      if (
        (intersect.size >= 2 && intersect.size === words1.size && intersect.size === words2.size) ||
        (intersect.size >= 3 && matchRatio >= 0.75) ||
        (intersect.size === 3 && words1.size === 3 && words2.size === 3)
      ) {
        matches.push(p2);
      }
    }

    if (matches.length > 1) {
      duplicates.push(matches);
      matches.forEach(m => processed.add(m.id));
    }
  }

  console.log(`Found ${duplicates.length} potential duplicate groups.`);
  duplicates.forEach((group, idx) => {
    console.log(`\nGroup ${idx + 1}:`);
    group.forEach((p: any) => console.log(`  - ${p.nombre} (ID: ${p.id}) [Esp: ${p.especialidad} | Otros: ${p.otros_cursos}]`));
  });
}

run();
